cpp/src/arrow/dataset/scanner.cc

```diff
@@ -182,7 +273,30 @@ Result<FragmentIterator> SyncScanner::GetFragments() {
   return GetFragmentsFromDatasets({dataset_}, scan_options_->filter);
 }
 
-Result<ScanTaskIterator> SyncScanner::Scan() {
+Result<ScanTaskIterator> SyncScanner::Scan() { return ScanInternal(); }
+
+Status SyncScanner::Scan(std::function<Status(TaggedRecordBatch)> visitor) {
+  ARROW_ASSIGN_OR_RAISE(auto scan_task_it, ScanInternal());
+
+  auto task_group = scan_options_->TaskGroup();
+
+  for (auto maybe_scan_task : scan_task_it) {
+    ARROW_ASSIGN_OR_RAISE(auto scan_task, maybe_scan_task);
+    task_group->Append([scan_task, visitor] {
+      ARROW_ASSIGN_OR_RAISE(auto batch_it, scan_task->Execute());
```

Comment:
I think with CSV this could lead to deadlock (https://github.com/apache/arrow/pull/9947#discussion_r611937558) since `Execute` (as opposed to `ExecuteAsync`) will call `MakeGeneratorIterator` which blocks on the next item (thus nested parallelism).  You could maybe come up with some way to work around it but at this point I feel like it would be much cleaner to revert out the async streaming CSV reader.  The patch added in ARROW-12161 isn't really compatible with where the new AsyncScanner ended up going either.  I can add it back in with ARROW-12355 as a mirror API instead of replacing the existing.

Reply:
Alright, let's go with the revert then. It does not revert cleanly so I'll put up a revert of ARROW-11887 and ARROW-12161 soon.

Reply:
Or actually, 11887 is probably okay and the only part is reverting file_csv.cc changes in 12161.

Reply:
See #10018

Reply:
Update: see #10019